### PR TITLE
TEP-0142: Referencing StepActions in Steps

### PR DESCRIPTION
**Pull request by chitrangpatel**
This PR allows the `Step` to reference a `StepAction` CRD deployed on the cluster. This capability is gated behind a feature flag `enable-step-actions`.  Remote resolution of `StepActions` will be implemented in a follow-up PR.

This is the fourth item on the implementation Issue https://github.com/tektoncd/pipeline/issues/7259

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
[WIP] Enables referencing of StepActions in Steps if the feature flag "enable-step-actions: true" is set.
```
/kind feature

### CODE REVIEW

**Comment by chitrangpatel** [comment]
@jerop I assume we also want to allow referencing via `v1beta1` api and have conversion functions between `v1` and `v1beta1`?

**Comment by jerop** [comment]
@tektoncd/core-maintainers this is ready for review, ptal 🙏🏾

File: pkg/reconciler/taskrun/resources/taskspec.go [hunk 1]
```diff
@@ -85,10 +91,53 @@ func GetTaskData(ctx context.Context, taskRun *v1.TaskRun, getTask GetTask) (*re
...
 	taskSpec.SetDefaults(ctx)
 	return &resolutionutil.ResolvedObjectMeta{
 		ObjectMeta:         &taskMeta,
 		RefSource:          refSource,
 		VerificationResult: verificationResult,
 	}, &taskSpec, nil
 }
+
+// extractStepActions extracts the StepActions and merges them with the inlined Step specification.
+func extractStepActions(ctx context.Context, taskSpec v1.TaskSpec, taskRun *v1.TaskRun, getStepAction GetStepAction) ([]v1.Step, error) {
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/taskspec.go:110]: is taskRun used in this function?

File: pkg/reconciler/taskrun/resources/taskspec.go [hunk 2]
```diff
@@ -85,10 +91,53 @@ func GetTaskData(ctx context.Context, taskRun *v1.TaskRun, getTask GetTask) (*re
...
+				return nil, apis.ErrGeneric("feature flag %s should be set to true to reference StepActions in Steps.", config.EnableStepActions)
+			}
+			s.Ref = nil
+			stepAction, _, err := getStepAction(ctx, step.Ref.Name)
+			if err != nil {
+				return nil, err
+			}
+			stepActionSpec := stepAction.StepActionSpec()
+			s.Image = stepActionSpec.Image
+			if stepActionSpec.Command != nil {
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/taskspec.go:125]: could it be that this is an empty list?

File: pkg/reconciler/taskrun/resources/taskspec_test.go [hunk 3]
```diff
@@ -286,11 +306,224 @@ func TestGetTaskData_VerificationResult(t *testing.T) {
...
+		want: &v1.TaskSpec{
+			Steps: []v1.Step{{
+				Image:  "myimage",
+				Script: "ls",
+			}},
+		},
+		stepActionFunc: func(ctx context.Context, n string) (*v1alpha1.StepAction, *v1.RefSource, error) {
+			stepAction := &v1alpha1.StepAction{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "stepActionWithSctipy",
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/taskspec_test.go:387]: this is a misspell?

File: pkg/reconciler/taskrun/resources/taskspec_test.go [hunk 4]
```diff
@@ -286,11 +306,224 @@ func TestGetTaskData_VerificationResult(t *testing.T) {
...
 }
+
+func TestGetStepAction(t *testing.T) {
+	tests := []struct {
+		name           string
+		tr             *v1.TaskRun
+		want           *v1.TaskSpec
+		stepActionFunc func(ctx context.Context, n string) (*v1alpha1.StepAction, *v1.RefSource, error)
+	}{{
+		name: "step-action-with-command",
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/taskspec_test.go:325]: do we miss args? I can only see command, script, env. And how about having them all in  1 case?

File: pkg/reconciler/taskrun/resources/taskspec.go [hunk 5]
```diff
@@ -85,10 +91,53 @@ func GetTaskData(ctx context.Context, taskRun *v1.TaskRun, getTask GetTask) (*re
...
 }
+
+// extractStepActions extracts the StepActions and merges them with the inlined Step specification.
+func extractStepActions(ctx context.Context, taskSpec v1.TaskSpec, taskRun *v1.TaskRun, getStepAction GetStepAction) ([]v1.Step, error) {
+	steps := []v1.Step{}
+	for _, step := range taskSpec.Steps {
+		s := step.DeepCopy()
+		if step.Ref != nil {
+			if !config.FromContextOrDefaults(ctx).FeatureFlags.EnableStepActions {
+				return nil, apis.ErrGeneric("feature flag %s should be set to true to reference StepActions in Steps.", config.EnableStepActions)
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/taskspec.go:116]: I don't see any code in reconciler returning apis.ErrGeneric, it is mainly used in webhooks. 
And do we need to validate this again here considering that we already have it in validation webhooks? And that validation function will also be invoked when creating the pod?

File: pkg/reconciler/taskrun/resources/taskspec_test.go [hunk 6]
```diff
@@ -286,11 +306,224 @@ func TestGetTaskData_VerificationResult(t *testing.T) {
...
+					Env: []corev1.EnvVar{{
+						Name:  "env1",
+						Value: "value1",
+					}},
+				},
+			}
+			return stepAction, nil, nil
+		},
+	}, {
+		name: "step-action-error",
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/taskspec_test.go:437]: this case can go to `TestGetStepAction_Error`

File: pkg/reconciler/taskrun/resources/taskspec.go (hunk 1 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec.go:110]: Good catch. I don't think it is. Thanks!

File: pkg/reconciler/taskrun/resources/taskspec.go (hunk 2 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec.go:125]: Yes, same for args. I think this is better: `if len(stepActionSpec.Command) > 0{...}`. I think this is what we use in validation as well.

File: pkg/reconciler/taskrun/resources/taskspec_test.go (hunk 3 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec_test.go:387]: yes. Fixed

File: pkg/reconciler/taskrun/resources/taskspec_test.go (hunk 4 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec_test.go:325]: We did not. That name was supposed to be "step-action-with-command-and-args". I can also unify them in one case.

Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec_test.go:325]: Actually I cannot since I cannot use command and script together.  so best to leave them separate.

File: pkg/reconciler/taskrun/resources/taskspec.go (hunk 5 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec.go:116]: Yes, I was just a bit paranoid. I can remove that check completely from the controller side.

File: pkg/reconciler/taskrun/resources/taskspec_test.go (hunk 6 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec_test.go:437]: agreed

File: pkg/reconciler/taskrun/resources/taskspec.go (hunk 2 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec.go]: fixed

File: pkg/reconciler/taskrun/resources/taskspec.go (hunk 1 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec.go]: fixed

File: pkg/reconciler/taskrun/resources/taskspec.go (hunk 5 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec.go]: fixed.

File: pkg/reconciler/taskrun/resources/taskspec_test.go (hunk 6 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskspec_test.go]: fixed.

File: pkg/reconciler/taskrun/resources/taskspec_test.go (hunk 4 above)
Comment (Yongxuanzhang) [reply pkg/reconciler/taskrun/resources/taskspec_test.go]: Oh yes, maybe 1 case for command, another one for script. and both have arge, env?

File: pkg/reconciler/taskrun/taskrun_test.go [hunk 7]
```diff
@@ -2734,6 +2735,121 @@ spec:
...
+					"enable-step-actions": "true",
+				},
+			},
+		},
+	}
+	testAssets, cancel := getTaskRunController(t, d)
+	defer cancel()
+	createServiceAccount(t, testAssets, "default", taskRun.Namespace)
+	c := testAssets.Controller
+	c.Reconciler.Reconcile(testAssets.Ctx, getRunName(taskRun))
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/taskrun_test.go:2838]: Is there an err returned here? If so can we check that err as well?

File: pkg/reconciler/taskrun/resources/taskref.go [hunk 8]
```diff
@@ -123,6 +123,15 @@ func GetTaskFunc(ctx context.Context, k8s kubernetes.Interface, tekton clientset
 	}
 }
 
+// GetStepActionFunc is a factory function that will use the given Ref as context to return a valid GetStepAction function.
+func GetStepActionFunc(tekton clientset.Interface, namespace string) GetStepAction {
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/taskref.go:127]: do we consider having a new file to store stepaction related functions? Not something we need to do right now

File: pkg/reconciler/taskrun/resources/taskref.go [hunk 9]
```diff
@@ -222,6 +231,26 @@ func (l *LocalTaskRefResolver) GetTask(ctx context.Context, name string) (*v1.Ta
 	return task, nil, nil, nil
 }
 
+// LocalRefResolver uses the current cluster to resolve a stepaction reference.
+type LocalRefResolver struct {
```
Comment (jerop) [inline pkg/reconciler/taskrun/resources/taskref.go:235]: just to make it specific, maybe down the road there'll be another `Ref`

```suggestion
type LocalStepActionRefResolver struct {
```

File: pkg/reconciler/taskrun/resources/taskref_test.go [hunk 10]
```diff
@@ -296,6 +309,104 @@ func TestLocalTaskRef(t *testing.T) {
...
+			}
+
+			if d := cmp.Diff(tc.expected, task); tc.expected != nil && d != "" {
+				t.Error(diff.PrintWantGot(d))
+			}
+
+			// local cluster step actions have empty source for now. This may be changed in future.
+			if refSource != nil {
+				t.Errorf("expected refsource is nil, but got %v", refSource)
+			}
```
Comment (jerop) [inline pkg/reconciler/taskrun/resources/taskref_test.go:406]: This logic is hard to follow, consider splitting this into two separate tests focused on valid and invalid cases instead of gating on `wantErr`

See this section of test [standards](https://github.com/tektoncd/community/blob/main/standards.md#tests):
> Table driven tests: do not use the same test for success and fail cases if the logic is different (e.g. do not have two sets of test logic, gated with `wantErr`)

Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskref_test.go:406]: Sure.

Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskref_test.go:406]: fixed.

File: pkg/reconciler/taskrun/resources/taskref.go (hunk 8 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/taskref.go:127]: May be. Lets see if the future PRs need further refactoring.

File: pkg/reconciler/taskrun/taskrun_test.go (hunk 7 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/taskrun_test.go:2838]: fixed! Thanks!

**Review by jerop — APPROVED** [review]
🎉 

thank you @chitrangpatel

**Review by jerop — COMMENTED** [review]
/lgtm
